python/tank/util/session.py

```diff
@@ -0,0 +1,181 @@
+from tank_vendor.shotgun_api3 import AuthenticationFault
+from ConfigParser import SafeConfigParser
+from .shotgun import get_associated_sg_config_data
+
+# Configure logging
+import logging
+logger = logging.getLogger("sgtk-session")
+# logger.setLevel(logging.DEBUG)
+logger.addHandler(logging.StreamHandler())
+
+
+def _get_cached_login_info_location(base_url):
+    """
+    Returns the location of the session file on disk for a specific site.
+    :param base_url: The site we want the login information for.
+    :returns: A dictionary with keys "login" and "session_token".
+    """
+    from tank.api import get_site_cache_root
+    return os.path.join(
+        get_site_cache_root(base_url),
+        "login.ini"
+    )
+
+
+def is_session_token_cached():
+    """
+    Returns if there is a cached session token for the current user.
+    :returns: True is there is, False otherwise.
+    """
+    if get_cached_login_info(get_associated_sg_config_data()["host"]):
+        return True
+    else:
+
+        return False
+
+
+def get_cached_login_info(base_url):
+    """
+    Returns the cached login info if found.
+    :returns: Returns a dictionnary with keys hostname, login and session_token or None
+    """
+    # Retrieve the location of the cached info
+    info_path = _get_cached_login_info_location(base_url)
+    # Nothing was cached, return an empty dictionary.
+    if not os.path.exists(info_path):
+        logger.debug("No cache found at %s", info_path)
+        return {}
+    try:
+        # Read the login information
+        config = SafeConfigParser({"login": None, "session_token": None})
+        config.read(info_path)
+        if not config.has_section("LoginInfo"):
+            logger.debug("No Login info was found")
+            return {}
+
+        login = config.get("LoginInfo", "login", raw=True)
+        session_token = config.get("LoginInfo", "session_token", raw=True)
+
+        if not login or not session_token:
+            logger.debug("Incomplete settings (login:%s, session_token:%s)", login, session_token)
+            return {}
+
+        return {"login": login, "session_token": session_token}
+    except Exception:
+        logger.exception("Exception thrown while loading cached session info.")
+        return {}
+
+
+def _validate_session_token(host, session_token, http_proxy):
+    """
+    Validates the session token by attempting to an authenticated request on the site.
+    :param session_token: Session token to use to connect to the host.
+    :param host: Host for that session
+    :param http_proxy: http_proxy to use to connect. If no proxy is required, provide None or an empty string.
+    """
+    # Connect to the site
+    logger.debug("Creating shotgun instance")
+    sg = Shotgun(
+        host.encode("utf-8"),
+        session_token=session_token.encode("utf-8"),
+        http_proxy=http_proxy.encode("utf-8") if http_proxy else None
+    )
+    try:
+        sg.find_one("HumanUser", [])
```

Comment:
Hmm, wondering if we should include some smarts in the shotgun API itself here? Let's discuss!